### PR TITLE
feat: support for multi keys

### PR DESCRIPTION
**Pull request by yhl25**

### CODE REVIEW

**Comment by ashwinidulams** [comment]
shuffle (when we get to that) should be immune to the ordering of keys.  conditional forwarding as well.

File: pkg/isb/testutils/udf.go [hunk 1]
```diff
@@ -33,13 +33,13 @@ func CopyUDFTestApply(ctx context.Context, readMessage *isb.ReadMessage) ([]*isb
 	_ = payload
 	// copy the payload
 	result := payload
-	var key string
+	var key []string
```
Comment (ashwinidulams) [inline pkg/isb/testutils/udf.go:36]: keys?

File: pkg/reduce/data_forward_test.go [hunk 2]
```diff
@@ -150,8 +150,8 @@ func (s SumReduceTest) ApplyReduce(_ context.Context, partitionID *partition.ID,
 	for msg := range messageStream {
 		var payload PayloadForTest
 		_ = json.Unmarshal(msg.Payload, &payload)
-		key := msg.Key
-		sums[key] += payload.Value
+		key := msg.Keys
```
Comment (ashwinidulams) [inline pkg/reduce/data_forward_test.go:153]:
```suggestion
		keys := msg.Keys
```

File: pkg/reduce/data_forward_test.go [hunk 3]
```diff
@@ -184,12 +184,12 @@ func (m MaxReduceTest) ApplyReduce(_ context.Context, partitionID *partition.ID,
 	for msg := range messageStream {
 		var payload PayloadForTest
 		_ = json.Unmarshal(msg.Payload, &payload)
-		if max, ok := maxMap[msg.Key]; ok {
+		if max, ok := maxMap[msg.Keys[0]]; ok {
```
Comment (ashwinidulams) [inline pkg/reduce/data_forward_test.go:187]: we should modify map to have []string as the key

Comment (yhl25) [reply pkg/reduce/data_forward_test.go:187]: the map cannot have []string as a key, so used the first element of the array in the tests.

Comment (ashwinidulams) [reply pkg/reduce/data_forward_test.go:187]: ah yes, these are slices.

**Review by ashwinidulams — APPROVED** [review]

**Review by whynowy — APPROVED** [review]
Nit - still some of `key -> keys` remaining.

File: pkg/isb/message.go [hunk 4]
```diff
@@ -59,8 +59,8 @@ type Header struct {
 	Kind MessageKind
 	// ID is used for exactly-once-semantics. ID is usually populated from the offset, if offset is available.
 	ID string
-	// Key is (key,value) in the map-reduce paradigm which will be used for conditional forwarding.
-	Key string
+	// Keys is (key,value) in the map-reduce paradigm which will be used for conditional forwarding.
```
Comment (vigith) [inline pkg/isb/message.go:62]: we need to fix this comment

File: pkg/sources/kafka/handler_test.go [hunk 5]
```diff
@@ -42,7 +42,7 @@ func TestMessageHandling(t *testing.T) {
 	partition := int32(1)
 	offset := int64(1)
 	value := "testvalue"
-	key := "testkey"
+	key := []string{"testkey"}
```
Comment (vigith) [inline pkg/sources/kafka/handler_test.go:45]:
```suggestion
	keys := []string{"testkey"}
```

File: pkg/sources/kafka/handler_test.go [hunk 6]
```diff
@@ -73,7 +73,7 @@ func TestMessageHandling(t *testing.T) {
 		Topic:     topic,
 		Partition: partition,
 		Offset:    offset,
-		Key:       []byte(key),
+		Key:       []byte(key[0]),
```
Comment (vigith) [inline pkg/sources/kafka/handler_test.go:76]:
```suggestion
		Key:       []byte(keys[0]),
```

File: pkg/sources/transformer/builtin/event_time/event_time_extractor.go [hunk 7]
```diff
@@ -54,7 +54,7 @@ func New(args map[string]string) (functionsdk.MapTFunc, error) {
 		format:     format,
 	}
 
-	return func(ctx context.Context, key string, datum functionsdk.Datum) functionsdk.MessageTs {
+	return func(ctx context.Context, key []string, datum functionsdk.Datum) functionsdk.MessageTs {
```
Comment (vigith) [inline pkg/sources/transformer/builtin/event_time/event_time_extractor.go:57]:
```suggestion
	return func(ctx context.Context, keys []string, datum functionsdk.Datum) functionsdk.MessageTs {
```

File: pkg/sources/transformer/builtin/filter/filter.go [hunk 8]
```diff
@@ -40,7 +40,7 @@ func New(args map[string]string) (functionsdk.MapTFunc, error) {
 		expression: expr,
 	}
 
-	return func(ctx context.Context, key string, datum functionsdk.Datum) functionsdk.MessageTs {
+	return func(ctx context.Context, key []string, datum functionsdk.Datum) functionsdk.MessageTs {
```
Comment (vigith) [inline pkg/sources/transformer/builtin/filter/filter.go:43]:
```suggestion
	return func(ctx context.Context, keys []string, datum functionsdk.Datum) functionsdk.MessageTs {
```

File: pkg/sources/transformer/builtin/filter/filter_test.go [hunk 9]
```diff
@@ -25,7 +25,7 @@ import (
 	"github.com/stretchr/testify/assert"
 )
 
-var _key = ""
+var _key = []string{""}
```
Comment (vigith) [inline pkg/sources/transformer/builtin/filter/filter_test.go:28]:
```suggestion
var _keys = []string{""}
```

File: pkg/sources/transformer/grpc_transformer.go [hunk 10]
```diff
@@ -81,18 +80,17 @@ func (u *gRPCBasedTransformer) WaitUntilReady(ctx context.Context) error {
 }
 
 func (u *gRPCBasedTransformer) ApplyMap(ctx context.Context, readMessage *isb.ReadMessage) ([]*isb.Message, error) {
-	key := readMessage.Key
+	key := readMessage.Keys
```
Comment (vigith) [inline pkg/sources/transformer/grpc_transformer.go:83]:
```suggestion
	keys := readMessage.Keys
```

File: pkg/sources/transformer/grpc_transformer.go [hunk 11]
```diff
@@ -81,18 +80,17 @@ func (u *gRPCBasedTransformer) WaitUntilReady(ctx context.Context) error {
...
 	payload := readMessage.Body.Payload
 	offset := readMessage.ReadOffset
 	parentMessageInfo := readMessage.MessageInfo
 	var d = &functionpb.Datum{
-		Key:       key,
+		Keys:      key,
```
Comment (vigith) [inline pkg/sources/transformer/grpc_transformer.go:88]:
```suggestion
		Keys:      keys,
```

File: pkg/sources/transformer/grpc_transformer.go [hunk 12]
```diff
@@ -107,7 +105,7 @@ func (u *gRPCBasedTransformer) ApplyMap(ctx context.Context, readMessage *isb.Re
 
 	writeMessages := make([]*isb.Message, 0)
 	for i, datum := range datumList {
-		key := datum.Key
+		key := datum.Keys
```
Comment (vigith) [inline pkg/sources/transformer/grpc_transformer.go:108]:
```suggestion
		keys := datum.Keys
```

File: pkg/sources/transformer/grpc_transformer.go [hunk 13]
```diff
@@ -116,7 +114,7 @@ func (u *gRPCBasedTransformer) ApplyMap(ctx context.Context, readMessage *isb.Re
 			Header: isb.Header{
 				MessageInfo: parentMessageInfo,
 				ID:          fmt.Sprintf("%s-%d", offset.String(), i),
-				Key:         key,
+				Keys:        key,
```
Comment (vigith) [inline pkg/sources/transformer/grpc_transformer.go:117]:
```suggestion
				Keys:        keys,
```

File: pkg/udf/builtin/filter/filter_test.go [hunk 14]
```diff
@@ -25,7 +25,7 @@ import (
 	"github.com/stretchr/testify/assert"
 )
 
-var _key = ""
+var _key = []string{""}
```
Comment (vigith) [inline pkg/udf/builtin/filter/filter_test.go:28]:
```suggestion
var _keys = []string{""}
```

File: test/fixtures/when.go [hunk 15]
```diff
@@ -41,7 +41,7 @@ type When struct {
 	kubeClient     kubernetes.Interface
 
 	portForwarderStopChannels map[string]chan struct{}
-	// Key: vertex label selector
+	// Keys: vertex label selector
```
Comment (vigith) [inline test/fixtures/when.go:44]:
```suggestion
	// Key: vertex label selector
```
this has nothing to do with keys in (keys, val) pair

**Review by vigith — COMMENTED** [review]
few more places to fix key -> keys and one revert

**Pull request merged**